### ISSUE TITLE
Training loses are nan with GPU, but with CPU it trains fine (with C3TR block)

### ISSUE THREAD

**Issue by rusvagzur** [ultralytics/ultralytics#1172]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

I am training [yolov8n-seg.yaml](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/v8/seg/yolov8n-seg.yaml) backbone but with a C3TR block in line 8 with my custom dataset. My custom dataset is fine, no corrupted labels or anything. When training with the yolov8n.yaml, everything works fine with GPU training. When training with transformer block C3TR and device='cpu', everything trains fine too. But including the C3TR module doesn't work when using GPU for training. I tried using coco128.yaml, different workers number, lower lr0, higher batch. To so sum up, I tried everything I found as an answer in other similar issues. I also tried v5loader=True. Nothing works. So my guess is that it's related to the GPU/CUDA....

### Environment

Ultralytics YOLOv8.0.46 🚀 Python-3.10.9 torch-1.13.1+cu117

My Ultralytics YOLO version is 8.0.46. I also tried from versions 8.0.41 up to 45 and the same happens.

My GPU is a GV100 My CUDA is 11.7. I also tried with CUDA 11.3. 

My Python is 3.10. I also tried Python 3.7, 3.8 and 3.9.

Pytorch 1.13.1, also tried Pytorch 1.12.1

Ubuntu 20.04

### Minimal Reproducible Example

backbone:
  - [-1, 1, Conv, [64, 3, 2]]  # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]]  # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]]  # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]]  # 7-P5/32
  - [-1, 3, C3TR [1024, False]] #<-------------------change introduced by me, from - [-1, 3, C2f [1024, True]]
  - [-1, 1, SPPF, [1024, 5]]  # 9

head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 13
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 17 (P3/8-small)
  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 20 (P4/16-medium)
  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  # 23 (P5/32-large)
  - [[15, 18, 21], 1, Segment, [nc, 32, 256]]  # Detect(P3, P4, P5)

Code for training:

`model.train(data="dataset.yaml", epochs=100, imgsz=320, name='yolov8n-seg-tr', project='transformer_block', save_period=1)`

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by rusvagzur** [comment]
https://github.com/ultralytics/ultralytics/issues/1348

bug was fixed in version 8.0.51

**Comment by rusvagzur** [comment]
@MathewsJosh You are actually right, this is issue is still not solved. I just dd the amp=False workaround in the trainer.py to make it work

**Comment by glenn-jocher** [comment]
@MathewsJosh @rusvagzur not sure what the cause is, but I'll add an amp argument to allow easier disabling, i.e. `yolo train amp=False`

**Comment by glenn-jocher** [comment]
@MathewsJosh @rusvagzur amp=True argument added in #1601 to allow you to easily disable AMP, i.e. yolo train amp=False or model.train(amp=False)

**Comment by JianDan0212** [comment]
Using this can indeed solve the problem of nan, but the parameters (P, R, mAP50, mAP50-95) are all 0, and the loss value of the training process has not converged

**Comment by pderrenger** [comment]
@JianDan0212 it sounds like the issue might be related to the AMP (Automatic Mixed Precision) configuration. When disabling AMP, it's important to adjust the learning rate accordingly as well, usually by multiplying it by 10x. This change can enable the model to converge properly and yield meaningful evaluation metrics. You can experiment with different learning rates and see if it improves the convergence of the training process.

**Comment by agentmorris** [comment]
@pderrenger You suggest increasing the learning rate when disabling AMP; is this still necessary if the optimizer is set to "auto", in which case console printouts tell us that lr0 will be determined automatically?  Or will the automatic parameter determination handle this internally?

Context: I have a model that was training fine with AMP on and a batch size of 16, but I re-started with a batch size of 32, and within a few epochs the val cls_loss became inf and showed no signs of recovering.  The fact that it was training fine with a smaller batch size suggests that there's nothing wrong with the data itself, but then I found this very helpful thread, disabled AMP, and things are working OK now.  GPU utilization and training speed were _really_ good when AMP was enabled and the batch size was higher; I haven't been able to achieve that utilization with the smaller batch size or with AMP disabled.  But, I can't always have my cake and eat it too, so I'll go with the larger batch size and no AMP.

**Comment by agentmorris** [comment]
Actually, I spoke too soon, with the higher batch size, disabling AMP pushes me just over my GPU memory capacity, and I eventually get a GPU OOM error.  I will stick with AMP and the batch size of 16.  Thanks in any case for this very helpful thread.
